quinn/src/tests.rs

```diff
@@ -183,70 +195,88 @@ fn export_keying_material() {
 }
 
 #[tokio::test]
-async fn accept_after_close() {
+async fn ip_blocking() {
     let _guard = subscribe();
-    let endpoint = endpoint();
-
-    const MSG: &[u8] = b"goodbye!";
-
-    let sender = endpoint
-        .connect(endpoint.local_addr().unwrap(), "localhost")
-        .unwrap()
-        .await
-        .expect("connect");
-    let mut s = sender.open_uni().await.unwrap();
-    s.write_all(MSG).await.unwrap();
-    s.finish().await.unwrap();
-    sender.close(0u32.into(), b"");
-
-    // Allow some time for the close to be sent and processed
-    tokio::time::sleep(Duration::from_millis(100)).await;
-
-    // Despite the connection having closed, we should be able to accept it...
-    let receiver = endpoint
-        .accept()
-        .await
-        .expect("endpoint")
-        .await
-        .expect("connection");
-
-    // ...and read what was sent.
-    let mut stream = receiver.accept_uni().await.expect("incoming streams");
-    let msg = stream
-        .read_to_end(usize::max_value())
-        .await
-        .expect("read_to_end");
-    assert_eq!(msg, MSG);
-
-    // But it's still definitely closed.
-    assert!(receiver.open_uni().await.is_err());
+    let endpoint_factory = EndpointFactory::new();
+    let client_1 = endpoint_factory.endpoint();
+    let client_1_addr = client_1.local_addr().unwrap();
+    let client_2 = endpoint_factory.endpoint();
+    let server = endpoint_factory.endpoint();
+    let server_addr = server.local_addr().unwrap();
+    let server_task = tokio::spawn(async move {
+        loop {
+            let accepting = server.accept().await.unwrap();
+            if accepting.remote_address() == client_1_addr {
+                accepting.reject();
+            } else if accepting.remote_address_validated() {
+                accepting.await.expect("connection");
+            } else {
+                accepting.retry().unwrap();
+            }
+        }
+    });
+    tokio::join!(
+        async move {
+            client_1
+                .connect(server_addr, "localhost")
+                .unwrap()
+                .await
+                .expect_err("server should have blocked this");
```

Comment:
We should assert that we get the correct error. For example, receiving a timeout here would be a bug.

Reply:
Done.